### ISSUE TITLE
RFE: Private (bind) mounts for a service

### ISSUE THREAD

**Issue by gdamjan** [systemd/systemd#3439]
### Submission type
- [x ] Request for enhancement (RFE)
### systemd version the issue has been seen with

systemd 230
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN
### Used distribution

ArchLinux
### Description

It would be usefull for a severly constrained service to have a private mount (in my case a bind mount) of a directory. I don't want to have that in fstab and be a global mount. But that seems to not be currently possible. The last thing I tried was:

```
[Service]
NoNewPrivileges=yes
PrivateDevices=yes
PrivateTmp=yes
ProtectSystem=full
MountFlags=private
PermissionsStartOnly=yes
User=owncloud
Group=owncloud
WorkingDirectory=/home/owncloud/
ExecStartPre=/usr/bin/mount --bind /home/music /home/owncloud/data/gdamjan/files/music
ExecStart=/usr/bin/uwsgi --ini /etc/uwsgi/owncloud.ini
```

but it seems that the bind mount in ExecStartPre is executed in its own mount namespace and doesn't have effect on the uwsgi process started in ExecStart (an assumption based that there's no error, but the process namespace doesn't have the mount, checked with nsenter).
The other option would be for the process started in ExecStart to use mount, but then I would have to run the service as root and depend on uwsgi dropping privileges.

**Comment by poettering** [comment]
So far we tried to avoid this, under the assumption that remapping the runtime file system view of a service is really more container territory than service territory.

That said, maybe adding PrivateBindPath= or so might actually be OK, as long as it is marketed as something for making only minor modifications to the tree...

**Comment by gdamjan** [comment]
another option would be to run all sub-processes in the same namespace when MountFlags=private, but I imagine that will complicate how systemd starts services.

but I'd be able to run any mount command I want in StartPre

**Comment by gdamjan** [comment]
How about .mount units suporting JoinNamespaceOf= ?

**Comment by gdamjan** [comment]
just as a followup, my (kind of) workaround for this requirement is to use `nsenter` in a privileged `ExecStartPost=+`

```
Type=notify
User=nextcloud
Group=nextcloud
ExecStart=/usr/bin/uwsgi --ini /etc/uwsgi/nextcloud.ini
ExecStartPost=+/usr/bin/nsenter -t $MAINPID -m /bin/mount --bind /some/aditional/path /home/nextcloud/data/user/files/xxx
PrivateTmp=yes
ProtectSystem=full
MountFlags=private
```

it might still make sense for systemd to support this in a declarative way, but at least I can start an unprivileged service with its own mounts like this too.

**Comment by gdamjan** [comment]
ps.
running the mount in `ExecStartPost=` is a bit racy, since the service is started before the mount is available. For this particular application that's just fine, but it might not be for other applications that would need the private mount to run at all.

**Comment by poettering** [comment]
Fix waiting in #4727
